File: 백준/Gold/6593. 상범 빌딩/상범 빌딩.py
import sys

# sys.stdin = open('input.txt', 'r')

dx = [-1, 1, 0, 0, 0, 0]
dy = [0, 0, -1, 1, 0, 0]
dz = [0, 0, 0, 0, -1, 1]

while True:
    l, r, c = map(int, input().split())
    if l == 0 and r == 0 and c == 0:
        break

    building = []
    start = end = None

    for z in range(l):
        floor = []
        while True:
            line = input().strip()
            if line == '':
                continue  # 빈 줄은 무시
            floor.append(list(line))
            if len(floor) == r:
                break
        building.append(floor)

    for z in range(l):
        for i in range(r):
            for j in range(c):
                if building[z][i][j] == 'S':
                    start = (z, i, j)
                elif building[z][i][j] == 'E':
                    end = (z, i, j)
    from collections import deque
    def bfs(start):
        visited = [[[-1] * c for _ in range(r)] for _ in range(l)]
        q = deque()
        q.append(start)
        visited[start[0]][start[1]][start[2]] = 0

        while q:
            z,x,y = q.popleft()

            if (z,x,y) == end:
                return visited[z][x][y]

            for d in range(6):
                nz = z + dz[d]
                ny = y + dy[d]
                nx = x + dx[d]

                if 0 <= nz < l and 0 <= nx < r and 0 <= ny < c:
                    if visited[nz][nx][ny] == -1 and building[nz][nx][ny] != '#':
                        visited[nz][nx][ny] = visited[z][x][y] + 1
                        q.append((nz,nx,ny))
        return - 1

    result = bfs(start)
    if result == -1:
        print("Trapped!")
    else:
        print(f"Escaped in {result} minute(s).")
        
    # 케이스 간 공백처리
    temp = input().strip()
    if temp == '':
        continue

File: 백준/Gold/6593. 상범 빌딩/상범 빌딩.js
const input = require('fs')
	.readFileSync('/dev/stdin', 'utf-8')
	.trim()
	.split('\n');

const directions = [
	[0, -1, 0],
	[0, 1, 0], // x축 (상하)
	[0, 0, -1],
	[0, 0, 1], // y축 (좌우)
	[-1, 0, 0],
	[1, 0, 0], // z축 (층 위아래)
];

let lineIndex = 0;

function isInBoundary(z, x, y, l, r, c) {
	return z >= 0 && z < l && x >= 0 && x < r && y >= 0 && y < c;
}

function bfs(building, start, end, l, r, c) {
	const visited = Array.from({ length: l }, () =>
		Array.from({ length: r }, () => Array(c).fill(false))
	);
	const [sz, sx, sy] = start;
	const [ez, ex, ey] = end;

	const queue = [];
	queue.push([...start, 0]);
	visited[sz][sx][sy] = true;

	while (queue.length > 0) {
		const [z, x, y, time] = queue.shift();

		if (z === ez && x === ex && y === ey) {
			return `Escaped in ${time} minute(s).`;
		}

		for (const [dz, dx, dy] of directions) {
			const [nz, nx, ny] = [z + dz, x + dx, y + dy];

			if (
				isInBoundary(nz, nx, ny, l, r, c) &&
				!visited[nz][nx][ny] &&
				building[nz][nx][ny] !== '#'
			) {
				visited[nz][nx][ny] = true;
				queue.push([nz, nx, ny, time + 1]);
			}
		}
	}

	return 'Trapped!';
}

function parseBuilding(l, r, c) {
	const building = [];
	let start = null;
	let end = null;

	for (let z = 0; z < l; z++) {
		const floor = [];
		for (let x = 0; x < r; x++) {
			const row = input[lineIndex++].split('');
			row.forEach((cell, y) => {
				if (cell === 'S') start = [z, x, y];
				if (cell === 'E') end = [z, x, y];
			});
			floor.push(row);
		}
		building.push(floor);
		lineIndex++;
	}

	return { building, start, end };
}
while (true) {
	const [l, r, c] = input[lineIndex++].split(' ').map(Number);

	if (l === 0 && r === 0 && c === 0) break;

	const { building, start, end } = parseBuilding(l, r, c);
	const result = bfs(building, start, end, l, r, c);
	console.log(result);
}
